apps/prairielearn/src/middlewares/authzCourseOrInstance.js

```diff
@@ -94,6 +95,27 @@ module.exports = asyncHandler(async (req, res, next) => {
     res.locals.authz_data.has_student_access_with_enrollment =
       permissions_course_instance.has_student_access_with_enrollment;
     res.locals.authz_data.has_student_access = permissions_course_instance.has_student_access;
+
+    // If the user does not currently have access to the course, but could if
+    // they were enrolled, automatically enroll them. However, we will not
+    // attempt to enroll them if they are an instructor (that is, if they have
+    // a specific role in the course or course instance).
+    if (
+      res.locals.authz_data.course_role === 'None' &&
+      res.locals.authz_data.course_instance_role === 'None' &&
+      res.locals.authz_data.has_student_access &&
+      !res.locals.authz_data.has_student_access_with_enrollment
+    ) {
+      // TODO: this enrollment should enforce enrollment limits.
+      await insertEnrollment({
+        user_id: res.locals.authn_user.user_id,
+        course_instance_id: res.locals.course_instance.id,
+      });
+
+      // Redirect them back to the same page/verb to pick up the new enrollment.
+      res.redirect(307, req.originalUrl);
```

Comment:
Alternatively, we could just set `res.locals.authz_data.has_student_access_with_enrollment = true` here? I think that's the only observable part of the authz data that would change. Originally, I wasn't sure about that, so I opted for the safer but slower "just repeat the entire request" approach.

Reply:
Cool, I updated that behavior in 109c41c.